### PR TITLE
refactor nms config

### PR DESCRIPTION
**Pull request by jshilong**
This PR refactored the code related to NMS. There are three specific points:

- 1 Remove the `nms_across_levels`  in all configs of  `RPNHead`  except for `GARPNHead`, because this argument is not used in  ``RPNHead``.

- 2 Rename the `max_num`  to `max_per_img` in configs of the whole types of `RPNHead`, to be consistent with the corresponding arguments in `RCNN`and  `One-Stage Detector`.

- 3 Rename the `nms_thr` to  `iou_threshold` and move it to a dict named `nms`  in configs of the whole type of `RPNHead`,  to be consistent with the way to specify the `nms` in `RCNN`and  `One-Stage Detector`.

## Migration

In the original implementation, We can't specify the `type` of NMS in `RPNHead` and use special argument names (`max_num` and `nms_thr`) compared to `RCNN` and `One-Stage Detector`.

For example, 

```python
train_cfg:
        rpn_proposal=dict(
            nms_across_levels=False,
            nms_pre=2000,
            nms_post=1000,
            max_num=1000,
            nms_thr=0.7,
            min_bbox_size=0),
            
            
test_cfg:
        rpn=dict(
            nms_across_levels=False,
            nms_pre=6000,
            nms_post=1000,
            max_num=1000,
            nms_thr=0.7,
            min_bbox_size=0),
        rcnn=dict(
            score_thr=0.05,
            nms=dict(type='nms', iou_threshold=0.5),
            max_per_img=100)
```

After this PR, It will be

```
train_cfg:
        rpn_proposal=dict(
            nms_pre=2000,
            nms_post=1000, # Only work in  `GARPNHead`
            max_per_img=1000,
            nms=dict(type='nms', iou_threshold=0.7),
            min_bbox_size=0),
            
            
test_cfg:
        rpn=dict(
            nms_pre=6000,
            nms_post=1000, # Only work in  `GARPNHead`
            max_per_img=1000,
            nms=dict(type='nms', iou_threshold=0.7),
            min_bbox_size=0),
        rcnn=dict(
            score_thr=0.05,
            nms=dict(type='nms', iou_threshold=0.5),
            max_per_img=100)
```

We can specify the type of NMS(such soft nms) in `rpn`. 

## BC-breaking

I have done compatibility to ensure the original config can be used.

### CODE REVIEW

File: mmdet/models/dense_heads/ga_rpn_head.py [hunk 1]
```diff
@@ -69,6 +69,15 @@ def _get_bboxes_single(self,
...
                            rescale=False):
         cfg = self.test_cfg if cfg is None else cfg
+
+        # refactor the nms cfg
+        # this is used for avoid breaking change
+        if 'nms' not in cfg:
+            cfg.nms = dict(type='nms', iou_threshold=cfg.nms_thr)
+            cfg.max_per_img = cfg.max_num
+
+        assert cfg.nms.get('type', 'nms') == 'nms', 'GARPNHead only support ' \
```
Comment (hhaAndroid) [inline mmdet/models/dense_heads/ga_rpn_head.py:106]: I think it’s strange, whether `assert cfg.nms.type == 'nms'` would be better.

File: mmdet/core/post_processing/merge_augs.py [hunk 2]
```diff
@@ -19,11 +23,40 @@ def merge_aug_proposals(aug_proposals, img_metas, rpn_test_cfg):
...
+    # deprecate arguments warning
+    if 'nms' not in cfg or 'max_num' in cfg or 'nms_thr' in cfg:
+        warnings.warn(
+            'In rpn_proposal or test_cfg, '
+            'nms_thr has been moved to a dict named nms as'
+            'iou_threshold, max_num has been renamed as max_per_img, '
+            'name of original arguments and the way to specify '
+            'iou_threshold of NMS will be deprecated.')
+    if 'nms' not in cfg:
+        cfg.nms = Config(dict(type='nms', iou_threshold=cfg.nms_thr))
```
Comment (hhaAndroid) [inline mmdet/core/post_processing/merge_augs.py:43]: Which is more appropriate, `Config` or `ConfigDict`?

File: mmdet/models/dense_heads/ga_rpn_head.py (hunk 1 above)
Comment (jshilong) [reply mmdet/models/dense_heads/ga_rpn_head.py:106]: We support to not specify the `type` in `cfg.nms` and use naive nms in this case.

File: mmdet/core/post_processing/merge_augs.py (hunk 2 above)
Comment (jshilong) [reply mmdet/core/post_processing/merge_augs.py:43]: sure, `ConfigDict` would be better.

**Review by hhaAndroid — APPROVED** [review]

**Pull request merged**